### PR TITLE
Feature/map : HeatMap improvements & Meeting requests

### PR DESCRIPTION
**Pull request by JanStaszewiczEPFL**
### Pull Request Description

**Summary:** Add Meeting Request Markers and Heatmap Toggle Functionality

**Description:**

This PR introduces two significant features to the map interface of the application:

1. **Meeting Request Markers:**
   - Implemented functionality to display markers on the map for users who have accepted meeting requests.
   - Each marker represents a user and is positioned based on their geographical coordinates.
   - Usernames are displayed above the markers, with a maximum character limit. If a username exceeds this limit, it is cropped and appended with "..." for better readability.
   - Markers are dynamically fetched based on the user's meeting requests, ensuring that the displayed information is always up-to-date.
   - Once you click on the marker it redirects to the user profile.
   - The pin disappears if the user is set as "already met"

2. **Heatmap Toggle:**
   - Added a toggle button to show or hide the heatmap overlay on the map.
   - The heatmap visualizes user locations based on their activity, providing insights into user density in specific areas.
   - The toggle functionality enhances user experience by allowing users to switch between viewing individual markers and the heatmap as needed.

**Changes Made:**
- Refractored `Heatmap` to `map`
- Updated `MapScreen` composable to include logic for fetching and displaying meeting request markers.
- Introduced a utility function to crop usernames for display purposes.
- Added a button to toggle the visibility of the heatmap overlay.
- Ensured that the overlay positions for markers are updated immediately upon rendering.

**Testing:**
- Still looking for a way to test what is displayed on the map.

**Screenshots:**
Screenshot_20241205_191612

### CODE REVIEW

File: app/src/main/java/com/github/se/icebreakrr/ui/map/MapView.kt [hunk 1]
```diff
@@ -80,18 +99,59 @@ fun HeatMap(
...
+      myProfile.value?.meetingRequestChosenLocalisation ?: emptyMap()
+
+  // Create a list of UIDs to fetch profiles
+  val uidsToFetch = meetingRequestChosenLocalisation.keys.toList()
+
+  // Create a mutable list to hold the fetched profiles
+  val profilesMeeting = remember { mutableStateListOf<Profile>() }
+
+  // Fetch profiles for the UIDs
+  LaunchedEffect(uidsToFetch) {
```
Comment (arthur-mrgt) [inline app/src/main/java/com/github/se/icebreakrr/ui/map/MapView.kt:115]: Problem: mutableStateListOf is used for profilesMeeting, which can lead to memory bloat if not cleaned up properly.
Solution: Clear profilesMeeting when it’s no longer needed

File: app/src/main/java/com/github/se/icebreakrr/ui/map/MapView.kt [hunk 2]
```diff
@@ -199,3 +304,37 @@ fun HeatMap(
...
+                IntOffset(
+                    position.x.roundToInt() - (overlayWidth.toPx() / 2).roundToInt(),
+                    position.y.roundToInt() + 10)
+              } // Center under the pin
+              .size(overlayWidth, overlayHeight) // Set the size of the overlay
+              .background(
+                  Color.Black.copy(alpha = 0.5f),
+                  shape =
+                      RoundedCornerShape(
+                          20.dp)) // Black background with transparency and rounded corners
```
Comment (arthur-mrgt) [inline app/src/main/java/com/github/se/icebreakrr/ui/map/MapView.kt:327]: Nitpick: Remove magic number

File: app/src/main/java/com/github/se/icebreakrr/ui/map/MapView.kt [hunk 3]
```diff
@@ -199,3 +304,37 @@ fun HeatMap(
...
+      ) {
+        Text(
+            text = text,
+            color = Color.White, // White text for contrast
+            fontSize = 9.sp, // Change this value to adjust the font size
```
Comment (arthur-mrgt) [inline app/src/main/java/com/github/se/icebreakrr/ui/map/MapView.kt:332]: Nitpick: Remove magic number

File: app/src/main/java/com/github/se/icebreakrr/ui/map/MapView.kt [hunk 4]
```diff
@@ -147,12 +208,56 @@ fun HeatMap(
...
 
+          userMarkers.forEach { userMarker ->
+            // Use the MarkerOverlay composable to display the text above the marker
+            userMarker.overlayPosition?.let {
+              MarkerOverlay(position = it, text = userMarker.username)
+            }
+          }
+
+          // Toggle button for heatmap visibility
+          Box(modifier = Modifier.padding(16.dp)) {
```
Comment (arthur-mrgt) [inline app/src/main/java/com/github/se/icebreakrr/ui/map/MapView.kt:255]: Nitpick: Remove magic number

File: app/src/main/java/com/github/se/icebreakrr/ui/map/MapView.kt [hunk 5]
```diff
@@ -80,18 +99,59 @@ fun HeatMap(
...
+
+  // Create UserMarkers from meetingRequestChosenLocalisation
+  val userMarkers =
+      meetingRequestChosenLocalisation.mapNotNull { (uid, pair) ->
+        val (message, coordinates) = pair
+        val profile = profilesMeeting.find { it.uid == uid }
+        profile?.let {
+          UserMarker(
+              uid = uid,
+              username = cropUsername(it.name, 10), // Crop username to 10 characters
```
Comment (arthur-mrgt) [inline app/src/main/java/com/github/se/icebreakrr/ui/map/MapView.kt:134]: Nitpick: Remove magic number

File: app/src/test/java/com/github/se/icebreakrr/utils/ProfileUtilsTest.kt
Comment (arthur-mrgt) [inline app/src/test/java/com/github/se/icebreakrr/utils/ProfileUtilsTest.kt:1]: Add Edge Cases, include edge cases such as:

- Negative maxLength.
- Usernames with special characters or whitespace.
- Very large usernames or max lengths.

Improve Readability: test names could describe the expected behavior in more detail.

**Review by arthur-mrgt — APPROVED** [review]
The code is excellent—clear, well-structured, and thoroughly tested. I only have a few minor suggestions:

Replace magic numbers (e.g., 10, 5, 0) with named constants for better readability and maintainability.
Expand test coverage with edge cases, such as negative maxLength, special characters, and very large inputs.
Consider adding a more explicit name for test to improve clarity on expected behavior.
Overall, great work—just a few tweaks to make it even more robust!

**Review by arthur-mrgt — CHANGES_REQUESTED** [review]
Implement the requested changes

File: app/src/main/java/com/github/se/icebreakrr/ui/map/MapView.kt (hunk 1 above)
Comment (JanStaszewiczEPFL) [reply app/src/main/java/com/github/se/icebreakrr/ui/map/MapView.kt:115]: You are right, I'll add some profiles cleaning in the code!

File: app/src/main/java/com/github/se/icebreakrr/ui/map/MapView.kt [hunk 6]
```diff
@@ -199,3 +304,37 @@ fun HeatMap(
...
+      ) {
+        Text(
+            text = text,
+            color = Color.White, // White text for contrast
+            fontSize = 9.sp, // Change this value to adjust the font size
+            fontWeight = FontWeight.Bold,
+            modifier =
+                Modifier.align(Alignment.Center) // Center the text within the box
+                    .wrapContentSize() // Ensure the text wraps correctly
+                    .padding(2.dp) // Optional: Add padding to the text for better spacing
```
[the remark this answers is not in the document]
Comment (JanStaszewiczEPFL) [reply app/src/main/java/com/github/se/icebreakrr/ui/map/MapView.kt:337]: yep thanks

File: app/src/main/java/com/github/se/icebreakrr/ui/map/MapView.kt (hunk 2 above)
Comment (JanStaszewiczEPFL) [reply app/src/main/java/com/github/se/icebreakrr/ui/map/MapView.kt:327]: yep yep yep

File: app/src/main/java/com/github/se/icebreakrr/ui/map/MapView.kt (hunk 3 above)
Comment (JanStaszewiczEPFL) [reply app/src/main/java/com/github/se/icebreakrr/ui/map/MapView.kt:332]: Wow again? I'll change that!

File: app/src/main/java/com/github/se/icebreakrr/ui/map/MapView.kt (hunk 4 above)
Comment (JanStaszewiczEPFL) [reply app/src/main/java/com/github/se/icebreakrr/ui/map/MapView.kt:255]: Let's remove those magic number!

File: app/src/main/java/com/github/se/icebreakrr/ui/map/MapView.kt (hunk 5 above)
Comment (JanStaszewiczEPFL) [reply app/src/main/java/com/github/se/icebreakrr/ui/map/MapView.kt:134]: That's a lot of magic numbers!

Comment (arthur-mrgt) [reply app/src/main/java/com/github/se/icebreakrr/ui/map/MapView.kt:144]: Indeed ;)

**Review by arthur-mrgt — APPROVED** [review]
Good fixes, slightly improved coverage. It's hard to test this class much more unfortunately. Good Job !
